src/main/java/edu/rpi/legup/ui/proofeditorui/treeview/TreeView.java

```diff
@@ -1,756 +1,881 @@
+            nodeView.draw(graphics2D);
+            for (TreeTransitionView transitionView : nodeView.getChildrenViews()) {
+                transitionView.draw(graphics2D);
+                redrawTree(graphics2D, transitionView.getChildView());
+            }
+        }
+    }
+
+    public void removeTreeElement(TreeElementView view) {
+        if (view.getType() == NODE) {
+            TreeNodeView nodeView = (TreeNodeView) view;
+            nodeView.getParentView().setChildView(null);
+        }
+        else {
+            TreeTransitionView transitionView = (TreeTransitionView) view;
+            transitionView.getParentViews().forEach((TreeNodeView n) -> n.removeChildrenView(transitionView));
+        }
+    }
+
+    /**
+     * When the edu.rpi.legup.user hovers over the transition, draws the corresponding rules image
+     *
+     * @param g the graphics to use to draw
+     */
+    public void drawMouseOver(Graphics2D g) {
+        if (selection.getHover().getType() == TRANSITION && ((TreeTransitionView) selection.getHover()).getTreeElement().isJustified()) {
+            TreeTransition transition = (TreeTransition) selection.getHover().treeElement;
+            int imgWidth = 100;
+            int imgHeight = 100;
+
+            BufferedImage image = new BufferedImage(imgWidth, imgHeight, BufferedImage.TYPE_INT_ARGB);
+            image.createGraphics().drawImage(transition.getRule().getImageIcon().getImage(), 0, 0, null);
+            Point mousePoint = selection.getMousePoint();
+            g.drawImage(image, mousePoint.x, mousePoint.y - 50, imgWidth, imgHeight, null);
+        }
+    }
+
+    public void resetView() {
+        this.tree = null;
+        this.rootNodeView = null;
+        this.selection.clearSelection();
+        this.selection.clearHover();
+    }
+
+    /**
+     * Called when a tree puzzleElement is added to the tree
+     *
+     * @param treeElement TreeElement that was added to the tree
+     */
+    @Override
+    public void onTreeElementAdded(TreeElement treeElement) {
+        if (treeElement.getType() == NODE) {
+            addTreeNode((TreeNode) treeElement);
+        }
+        else {
+            addTreeTransition((TreeTransition) treeElement);
+        }
+        repaint();
+    }
+
+    /**
+     * Called when a tree puzzleElement is removed from the tree
+     *
+     * @param element TreeElement that was removed to the tree
+     */
+    @Override
+    public void onTreeElementRemoved(TreeElement element) {
+        if (element.getType() == NODE) {
+            TreeNode node = (TreeNode) element;
+            TreeNodeView nodeView = (TreeNodeView) viewMap.get(node);
+
+            nodeView.getParentView().setChildView(null);
+            removeTreeNode(node);
+        }
+        else {
+            TreeTransition trans = (TreeTransition) element;
+            TreeTransitionView transView = (TreeTransitionView) viewMap.get(trans);
+
+            //unlock ancestor elements if case rule deleted
```

Comment:
Space between the comment sequence and the content itself.

```java
//unlock ancestor elements if case rule deleted
```
should be
```java
// unlock ancestor elements if case rule detected
```

Reply:
added